chain-signatures/node/src/mesh/connection.rs

```diff
@@ -1,245 +1,221 @@
-                }
-                Ok(Err(())) => {
-                    // Already logged in task
-                }
-                Err(e) => {
-                    tracing::warn!("fetch participant state task panicked: {e}");
-                }
-            }
+impl Pool {
+    pub fn new(client: &NodeClient, ping_interval: Duration) -> Self {
+        tracing::info!(target: "net[pool]", "creating new connection pool");
+        Self {
+            client: client.clone(),
+            ping_interval,
+            connections: HashMap::new(),
         }
-
-        drop(status);
-
-        // Update the active participants
-        let mut potential_active = self.potential_active.write().await;
-        *potential_active = Some((participants.clone(), Instant::now()));
-
-        participants
     }
 
-    pub async fn establish_participants(&self, contract_state: &ProtocolState) {
-        match contract_state {
-            ProtocolState::Initializing(contract_state) => {
-                let participants: Participants = contract_state.candidates.clone().into();
-                self.set_participants(&participants).await;
+    pub async fn connect(&mut self, contract: &ProtocolState) {
+        let mut seen = HashSet::new();
+        match contract {
+            ProtocolState::Initializing(init) => {
+                let participants: Participants = init.candidates.clone().into();
+                self.connect_nodes(&participants, &mut seen).await;
             }
-            ProtocolState::Running(contract_state) => {
-                self.set_participants(&contract_state.participants).await;
+            ProtocolState::Running(running) => {
+                self.connect_nodes(&running.participants, &mut seen).await;
             }
-            ProtocolState::Resharing(contract_state) => {
-                self.set_participants(&contract_state.old_participants)
-                    .await;
-                self.set_potential_participants(&contract_state.new_participants)
+            ProtocolState::Resharing(resharing) => {
+                // NOTE: do NOT connect with old participants since only the new ones are
+                // operating under the new epoch and talking to each other. In the case of
+                // a resharing revert, we will go back to running state from the contract,
+                // and then the old participants would be connected again.
+                self.connect_nodes(&resharing.new_participants, &mut seen)
                     .await;
             }
         }
-        tracing::debug!(
-            "Pool.establish_participants set participants to {:?}",
-            self.connections.read().await.clone().keys_vec()
-        );
-    }
 
-    async fn set_participants(&self, participants: &Participants) {
-        *self.connections.write().await = participants.clone();
+        // drop the connections that are not in the seen list
+        self.drop_connections(seen);
```

Comment:
Hm, without the comment, I would misunderstand the function signature and what it does. Whenever I need a comment like this, I take it as a sign to change the function name to make the comment superfluous.

How about `retain_connections` instead of `drop_connections`?